File: baekjoon/src/dfsAndBfs/No2667.java
package dfsAndBfs;

import java.io.*;
import java.util.Arrays;

/**
 * https://www.acmicpc.net/problem/2667
 */
public class No2667 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static int square;
    public static int[][] graph;
    public static int[] cities;
    public static int index;
    public static int city = 11;

    public static void setGraph() throws IOException {
        square = Integer.parseInt(reader.readLine());
        graph = new int[square][square];
        cities = new int[625];
        for (int i = 0; i < square; i++) {
            String cities = reader.readLine();
            for (int j = 0; j < square; j++)
                graph[i][j] = cities.charAt(j) - '0';
        }
    }

    public static void updateCities() {
        for (int i = 0; i < square; i++) {
            for (int j = 0; j < square; j++) {
                if (graph[i][j] == 1) {
                    bfs(i, j, city++);
                    index++;
                }
            }
        }
    }

    public static void bfs(int row, int col, int city) {
        graph[row][col] = city;
        cities[index]++;
        for (int i = row; i < square; i++) {
            for (int j = col; j < square; j++) {
                if (row - 1 >= 0 && graph[row - 1][col] == 1) {
                    graph[row - 1][col] = city;
                    bfs(row - 1, col, city);
                } else if (col + 1 < square && graph[row][col + 1] == 1) {
                    graph[row][col + 1] = city;
                    bfs(row, col + 1, city);
                } else if (row + 1 < square && graph[row + 1][col] == 1) {
                    graph[row + 1][col] = city;
                    bfs(row + 1, col, city);
                } else if (col - 1 >= 0 && graph[row][col - 1] == 1) {
                    graph[row][col - 1] = city;
                    bfs(row, col - 1, city);
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        setGraph();
        updateCities();

        writer.write(String.valueOf(index));
        writer.newLine();
        Integer[] arr = new Integer[index];
        for (int i = 0; i < index; i++)
            arr[i] = cities[i];
        Arrays.sort(arr);

        for (int i = 0; i < index; i++) {
            writer.write(String.valueOf(arr[i]));
            writer.newLine();
        }
        writer.flush();
    }
}

File: baekjoon-classification/src/dfs/No2667.kt
package dfs

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

/**
 * https://www.acmicpc.net/problem/2667
 */
private var size = 0
private var count = 0
private val dx = intArrayOf(1, 0, -1, 0)
private val dy = intArrayOf(0, -1, 0, 1)

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    size = readLine().toInt()
    val graph = Array(size) { BooleanArray(size) }
    repeat(size) { i ->
        val row = readLine()
        graph[i] = BooleanArray(size) { j -> row[j] == '1' }
    }

    val houses = arrayListOf<Int>()
    for (i in graph.indices) {
        for (j in graph.indices) {
            if (graph[i][j]) {
                count = 0
                dfs(graph, i, j)
                houses.add(count)
            }
        }
    }

    houses.sort()

    with(BufferedWriter(OutputStreamWriter(System.out))) {
        write("${houses.size}")
        newLine()
        for (house in houses) {
            write("$house")
            newLine()
        }
        flush()
    }
}

private fun dfs(graph: Array<BooleanArray>, row: Int, col: Int) {
    graph[row][col] = false
    count++
    for (i in 0 until 4) {
        val nextRow = row + dy[i]
        val nextCol = col + dx[i]
        if (isInBoundary(nextRow, nextCol) && graph[nextRow][nextCol])
            dfs(graph, nextRow, nextCol)
    }
}

private fun isInBoundary(row: Int, col: Int) = row in 0 until size && col in 0 until size